# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def add_two(x):
    return x+2

print(add_two(3))

#as a lambda function 

lambda x : x+2

print((lambda x: x+2)(3))

5
5


#### Saving to a Variable

In [3]:
f = lambda x, y: x+y+2

print((lambda x, y: x+y+2)(2,3))

print(f(2,3))

7
7


#### Multiple Inputs

In [4]:
(lambda x,y,z:x*y*z)(3,5,8)

120

#### Passing a Lambda into a Function

In [6]:
def do_math(f, x):
    return f(x)

do_math(lambda x: x*x, 7)

49

#### Returning a Lambda from a Function

In [8]:
def return_f():
    def multiply(x):
        return x * x
    return multiply
print(type(return_f()))

return_f()(4)



<class 'function'>


16

#### If Statements within Lambdas

In [10]:
#Lambda x : True if (condition) else false
f = lambda n: n*2 if n>10 else n+2

f(12)
f(10)


12

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [17]:
f = lambda x: x**3

f(2)
f(3)

27

In [18]:
#get good at reading Documentation because this is your lifelinealso
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [24]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest
a = {
    "name":"Joe",
    "age":52
}

b= {
    "name":"Jack",
    "age":71
}

c= {
    "name":"June",
    "age":8
}

d= {
    "name":"Fern",
    "age":63
}


p = [a,b,c,d]

print(p)

l = lambda p_dict: p_dict["age"]

print(sorted(p, key = l))

print(sorted(p, key = lambda p_dict: p_dict["age"], reverse = True))

[{'name': 'Joe', 'age': 52}, {'name': 'Jack', 'age': 71}, {'name': 'June', 'age': 8}, {'name': 'Fern', 'age': 63}]
[{'name': 'June', 'age': 8}, {'name': 'Joe', 'age': 52}, {'name': 'Fern', 'age': 63}, {'name': 'Jack', 'age': 71}]
[{'name': 'Jack', 'age': 71}, {'name': 'Fern', 'age': 63}, {'name': 'Joe', 'age': 52}, {'name': 'June', 'age': 8}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [25]:
# map(func,iterable(list,dict,tuple,set,etc...))
# Normally the usage of map happens with a pre-defined function -- but can use lambdas as well
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [28]:
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2
    
num=[4,10,5,4,2,]
num2=[4,11,20,3,15]

squared_nums_map = map(squared, num, num2)
print(list(squared_nums_map))

[(16, 16), (10, 11), (5, 20), (16, 9), (2, 15)]


#### Using Lambda's with Map

In [ ]:
# map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))


In [33]:
num=[4,10,5,4,2,]
squared = list(map((lambda x: x**2 if x < 10 else x), num))
print(squared)

[16, 10, 25, 16, 4]


In [1]:
num=[4,10,5,4,2,]
num2=[4,11,20,3,15]

squared_nums2 = list(map(
    (lambda x1, y1: (x1**2, y1**2) if x1<10 and y1<10 else (x1,y1)), num, num2))
squared_nums2

[(16, 16), (10, 11), (5, 20), (16, 9), (2, 15)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [2]:
nums = [4,2,6,8]

dub_minus_one = list(map((lambda x: x*2 - 1), nums))

print(dub_minus_one)

[7, 3, 11, 15]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [57]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [3]:
list(filter(None,(True, True, False, False, True)))

[True, True, True]

In [4]:
names = ['Anna', 'Alex', 'Beth', 'David', 'Daniel', 'Jamal', 'Kristina', 'Krystyna', 'David']
#we want names that start with K

def k_name(name): 
    if name[0].lower() == 'k':
        return True
    else:
        return False
    
new_names = list(filter(k_name, names))
print(new_names)

['Kristina', 'Krystyna']


#### Using Lambda's with Filter()

In [11]:
new_names = list(filter(lambda name: True if name[0].lower() == 'k' else False, names))
print(new_names)

['Kristina', 'Krystyna']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [18]:
from statistics import mean
num_list = [2,3,5,8]
new_num = list(filter(lambda num: True if num >= mean(num_list) else False, num_list))
print(new_num)

[5, 8]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [5]:
from functools import reduce
# reduce(function,iterable)
help(reduce)


Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [22]:
list = [2,4,6,8,10]

def sub(num1, num2):
    return num1 - num2

result = reduce(sub, list)
result

-26

#### Using Lambda's with Reduce()

In [21]:
list3 = [1,10,2,30]
result = reduce(lambda x,y: x-y,list3)
print(result)

-41


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [6]:
l5 = [1,2,3,4]

result = reduce(lambda x,y: x*y, l5)
print(result)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [7]:
def recur(i):
    print("hi")
    if i == 4:
        return
    else:
        i+=1
        recur(i)
    
recur(0)

hi
hi
hi
hi
hi


#### Implementing a Base Case

In [32]:
def add_nums (num):
    if num <= 1:
        return num
    else:
        x = num -1
        return num + add_nums(x)
    
add_nums(5)

15

#### Writing a Factorial Function

In [38]:
# 5! = 5*4*3*2*1
def fact(num):
    if num <= 1:
        return 1
    else:
        return num*fact(num-1)
fact(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [8]:
def rec_sub(num):
    if num <= 1:
        return(1)
    else:
        return num - rec_sub(num-1)
    
rec_sub(10)


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [9]:
def test():
    yield 1
    yield 2
    yield 3
    yield "a"
    yield ['a','b','c']
    
for i in test():
    print(i)

1
2
3
a
['a', 'b', 'c']


In [10]:
range(5)

range(0, 5)

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [14]:
def my_range(n):
    for i in range(n):
        yield i
list(my_range(5))

[0, 1, 2, 3, 4]

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [15]:
def square(stop, start, step=1):
    while start>stop:
        yield start**2
        start -= step 
    
for num in square(0,10,1):
    print(num)

100
81
64
49
36
25
16
9
4
1


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [22]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

places_filtered = list(filter(lambda place: False if place.strip() == "" else True, places))

print(places_filtered)

['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [19]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key= lambda name: name.lower().split()[-1])

print(author)

['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [27]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

list(map(lambda place: [place[0], place[1]*(9/5) + 32], places))

[['Nashua', 89.6], ['Boston', 53.6], ['Los Angelos', 111.2], ['Miami', 84.2]]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [83]:
# The Fibonacci sequence is a series of numbers in which each number is the sum of the two that 
# precede it. Starting at 0 and 1, the sequence looks like this: 0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 
# and so on forever. With this, iteration 0 should be 0, 1 should be 1, and 2 should be 1, etc


def fib(num, stop = 0, step=1):
    while num>stop:
        if num<1:
            yield num
        else:
            yield (my_range(num-1) + my_range(num-2))
            start -= step 
    
for n in fib(5):
    print(n)
    
have tried this at least 10 different ways and cannot get it. tried recursion, tried generator,
examples online mostly just appear to be generators, however I couldn't even get that to work.
giving up because I refuse to turn in work that is just copy/past from someone else. 

TypeError: unsupported operand type(s) for +: 'NoneType' and 'NoneType'